### PR TITLE
Add the example on turbulent taylor-couette flow

### PR DESCRIPTION
**Pull request by blaisb**
# Description of the problem

- Using the Matrix Free solver, Joel cooked up a very cool 3d Taylor-Couette example that really showcases the power of the matrix free approach

# Description of the solution

- To keep this example for the future, we add it as an example test case.

# How Has This Been Tested?

- Runs well, this is super chill.

### CODE REVIEW

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    rho = prm.rho
+    omega = prm.omega
+    ri = prm.ri
+    d = prm.d
+    U = omega*ri
+    
+    #Adimensionnal factor
+    coeff = (2*d*d)/(rho*U*U)
+    
+    #Recovering datas
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:31]:
```suggestion
    #Recovering data
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    file = file_name 
+    path = os.path.join(os.path.dirname(__file__), file)
+
+    #Stocking datas
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:35]:
```suggestion
    #Loading data
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+import numpy as np 
+import os
+
+
+def enstrophy (prm,file_name):
+    """Function that retrieves and returns, in the form of arrays, the time and adimensional kinetic energy values obtained during simulation.
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:7]:
```suggestion
    """Function that retrieves and returns, in the form of arrays, the time and dimensionless kinetic energy values obtained during simulation.  
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    """ 
+    
+    #Parameter variables
+    rho = prm.rho
+    omega = prm.omega
+    ri = prm.ri
+    d = prm.d
+    U = omega*ri
+    
+    #Adimensionnal factor
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:28]:
```suggestion
    #Dimensionless factor
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    
+def enstrophy_ref(): 
+    """Function that retrieves and returns, in the form of arrays, the time and kinetic energy values of references.  
+
+    Entrées :
+        - prm : Object class parameter () 
+            - ri = Radius of  the inner cylinder
+            - rho = Fluid density
+            - omega = Angular velocity of the inner cylidner 
+            - d = Annulus width (re-ri)
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:51]:
```suggestion
```

Comment (PierreLaurentinCS) [reply examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:51]: No inputs for this function

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        
+            
+    Sortie (Dans l'ordre énuméré ci-dessous) :
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:54]:
```suggestion
    Outputs (same order as below):
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        - t4 = Vector (np.array) of reference times (P4)
+        - E4 = Vector (np.array) of all reference dimensionless enstrophy values (P4)
+        - t5 = Vector (np.array) of reference times (P5)
+        - E5 = Vector (np.array) of all reference dimensionless enstrophy values (P5)
+    """ 
+    
+    #Recovering datas
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:61]:
```suggestion
    #Recovering data
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    
+    #Parameter variables
+    rho = prm.rho
+    omega = prm.omega
+    ri = prm.ri
+    d = prm.d
+    U = omega*ri
+    
+    #Adimensionnal factor
+    coeff = (2*d*d)/(rho*U*U)
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:29]: This dimensionless coefficient seems to have a dimension of s^2.m^3/kg

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    #Recovering datas
+    file3 = 'enstrophy_wang_p3.tsv'
+    file4 = 'enstrophy_wang_p4.tsv'
+    file5 = 'enstrophy_wang_p5.tsv'
+    dir = 'references'
+    path3 = os.path.join(os.path.dirname(__file__),dir,file3)
+    path4 = os.path.join(os.path.dirname(__file__),dir,file4)
+    path5 = os.path.join(os.path.dirname(__file__),dir,file5)
+    
+    #Stocking datas
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-functions.py:70]:
```suggestion
    #Loading data
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+# Simulation Control 
+#---------------------------------------------------
+
+subsection simulation control
+  set method            = bdf3
+  set output name       = taylor-couette
+  set output path       = ./output1/
+  set time end          = 60
+  set time step         = 0.0125
+  set number mesh adapt = 0
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:16]:
```suggestion
```

Comment (PierreLaurentinCS) [reply examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:16]: It's a default parameter

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  set pressure order = 2      #3 for Q3
+end
+
+#---------------------------------------------------
+# Post-Processing
+#---------------------------------------------------
+
+subsection post-processing
+  set verbosity                        = quiet
+  set output frequency                 = 1
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:81]:
```suggestion
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+  # Kinetic energy calculation
+  set calculate kinetic energy         = true
+  set kinetic energy name              = kinetic_energy
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:85]:
```suggestion
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+  # Enstrophy calculation
+  set calculate enstrophy              = true
+  set enstrophy name                   = enstrophy
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:89]:
```suggestion
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    set periodic_direction = 2
+  end
+end
+
+#---------------------------------------------------
+# Mesh Adaptation Control
+#---------------------------------------------------
+subsection mesh adaptation
+  set type = none
+end
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:127]:
```suggestion
```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-postprocessing.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    omega = 1.0       #Angular velocity (rad/s)
+    d = 0.5           #Width of the annulus (re - ri)
+    alpha = 2*np.pi   #Aspect ratio (L/d)
+    nu = 6.25e-5      #Kinematic viscosity
+    rho = 1.0         #Fluid density 
+    Re = 4000         #Reynolds number
+
+prm = parameter() 
+
+### Simulations datas ###
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-postprocessing.py:26]:
```suggestion
### Simulations data ###

```

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-postprocessing.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+#enstrophy
```
Comment (PierreLaurentinCS) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-postprocessing.py:27]:
```suggestion
#Enstrophy

```

**Review by PierreLaurentinCS — CHANGES_REQUESTED** [review]
Very cool example featuring the power of matrix-free Lethe. I've put minor comments here and there but overall it's a very good job!

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+    * - :math:`r_i`
+      - 0.5
+      - :math:`r_o`
+      - 1.0
+
+Since Lethe uses a Cartesian coordinate system, the following expressions have been transformed to proceed with the simulation. Finally, the histories of the enstrophy and kinetic energy are use as a benchmark. For an incompressible flow, the enstrophy and the kinetic energy are defined as: 
+
+.. math::
+  \mathcal{E} &= \frac{1}{\Omega} \int_{\Omega} \frac{\mathbf{\omega}\cdot \mathbf{\omega}}{2} \mathrm{d}\Omega \\
+  E_k &= \frac{1}{\Omega} \int_{\Omega} \frac{\mathbf{u}\cdot \mathbf{u}}{2} \mathrm{d}\Omega \\
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:83]: Yeah I understand, it's the kinetic energy per unit of mass and volume. It's the definition we use in the regular Lethe doc which I agree is maybe not the most adequate. If we change that I'd change it in another PR however.

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+      set time end          = 60  
+      set output frequency  = 2    
+      set subdivision       = 2
+    end
+
+.. tip::
+
+  A good practice is to use as many subdivisions as the interpolation order scheme. 
+
+Rest of the Subsections
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:257]: lolz

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+  python3 tc_postprocessing.py 
+
+The enstrophy plot features a zoom section of the enstrophy cascade. The following plot shows the history of the enstrophy as measured with the Q2 scheme : 
+
++-------------------------------------------------------------------------------------------------------------------+
+|  .. figure:: images/enstrophy_comparison_82k_Q2.png                                                               |
+|     :width: 620                                                                                                   |
+|                                                                                                                   |
++-------------------------------------------------------------------------------------------------------------------+
+
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:300]: I'll need to re-run the simulations which will be a bit difficult to do remotely for now, but I can keep this in mind and fix it in a future PR

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+    :name: geometry
+    :height: 6cm
+
+The initial conditions for velocity and pressure are defined as follow: 
+
+.. math::
+   u_{\theta} &= Ar + \frac{B}{r} + \epsilon U\sin(\theta) \sin(\frac{(r-r_i)\pi}{r_i}) \sin(\frac{z}{d}) \\
+   u_{r} &= \epsilon U\cos(\theta) \sin(\frac{(r-r_i)\pi}{r_i}) \sin(\frac{z}{d}) \\
+   u_{z} &= 0 \\
+   p &= \frac{1}{2}A^2r^2 + 2AB\ln(r) - \frac{\frac{1}{2}B^2}{r^2} + \frac{1}{2}(\epsilon U)^2 \cos(2\theta) \sin(\frac{2(r-r_i)\pi}{r_i}) \sin(\frac{2z}{d})
```
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:49]:
```suggestion
.. math::
   u_{\theta} &= Ar + \frac{B}{r} + \epsilon U\sin(\theta) \sin \left( \frac{(r-r_i)\pi}{r_i} \right) \sin \left( \frac{z}{d} \right) \\
   u_{r} &= \epsilon U\cos(\theta) \sin \left( \frac{(r-r_i)\pi}{r_i} \right) \sin \left( \frac{z}{d} \right) \\
   u_{z} &= 0 \\
   p &= \frac{1}{2}A^2r^2 + 2AB\ln(r) - \frac{\frac{1}{2}B^2}{r^2} + \frac{1}{2}(\epsilon U)^2 \cos(2\theta) \sin  \left( \frac{2(r-r_i)\pi}{r_i} \right) \sin \left( \frac{2z}{d} \right)
```

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+
+.. image:: images/mesh.png
+    :alt: The mesh
+    :align: center
+    :name: mesh_taylor_couette
+    :height: 6cm
+
+.. note::
+
+  The mesh resolution used in the case study consists of 20 radial elements, 80 azimuthal elements, and 64 axial elements totalling 102,400 elements. Additionally, it incorporates a mesh refinement near both walls in the radial direction.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:119]:
```suggestion
  The mesh resolution used in the case study consists of 20 radial elements, 80 azimuthal elements, and 64 axial elements totalling 102,400 elements.  
```

File: doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/3d-turbulent-taylor-couette/3d-turbulent-taylor-couette.rst:300]: I fixed this now

**Comment by blaisb** [comment]
I applied all of your comments @PierreLaurentinCS so if you are ok with it I will merge

**Review by PierreLaurentinCS — APPROVED** [review]
All good !

**Pull request merged**